lib/Dialect/TritonGPU/IR/Dialect.cpp

```diff
@@ -1210,6 +1288,417 @@ void BlockedEncodingAttr::print(mlir::AsmPrinter &printer) const {
+  return basesPerDim(getLinearLayout(), StringAttr::get(getContext(), "lane"));
+}
+SmallVector<unsigned> LinearEncodingAttr::getThreadOrder() const {
+  return orderPerDim(getLinearLayout(), StringAttr::get(getContext(), "lane"),
+                     getRepOrder__());
+}
+SmallVector<unsigned> LinearEncodingAttr::getSizePerThread() const {
+  auto rank = getRepOrder().size();
+  auto ll = getLinearLayout();
+  auto kRegister = StringAttr::get(getContext(), "register");
+  if (getCTAsPerCGA() == SmallVector<unsigned>(rank, 1)) {
+    return basesPerDim(ll, kRegister);
+  } else {
+    // We canonicalize on the spot, as if we use CGAs the regs are not in
+    // canonical form The order is [reg, lane, warp, rep, block], so we first
+    // remove the blocks
+    llvm::SmallVector<unsigned> ctaShape;
+    for (auto [shape, cgaNum] :
+         llvm::zip(ll.getOutDimSizes(), getCTASplitNum())) {
+      ctaShape.push_back(shape / cgaNum);
+    }
+    LinearLayout::BasesT bases = ll.getBases();
+    removeRepsInPlace(getContext(), bases, getRepOrder(), ctaShape);
+    return basesPerDim(bases, kRegister, rank);
+  }
+}
+SmallVector<unsigned> LinearEncodingAttr::getOrder() const {
+  return orderPerDim(getLinearLayout(),
+                     StringAttr::get(getContext(), "register"),
+                     getRepOrder__());
+}
+
+SmallVector<unsigned>
+LinearEncodingAttr::getShapePerCTATile(ArrayRef<int64_t> tensorShape) const {
+  // FIXME Remove this function from the Distributed API
+  // It's not entirely clear what this function should return (it doesn't use
+  // tensorShape!) We follow the definition in BlockedEncodingAttr, but we
+  // should just remove it
+  auto ll = getLinearLayout();
+  SmallVector<unsigned> shape;
+  for (unsigned d = 0, n = getRepOrder().size(); d < n; ++d) {
+    shape.push_back(getSizePerThread()[d] * getThreadsPerWarp()[d] *
+                    getWarpsPerCTA()[d]);
+  }
+  return shape;
+}
+
+std::optional<LinearLayout>
+LinearEncodingAttr::toLinearLayout(ArrayRef<int64_t> shape) const {
+  auto ll = getLinearLayout();
+  auto canonicalDims = llvm::to_vector(ll.getOutDimNames());
+  llvm::SmallDenseMap<StringAttr, int64_t> namedShape;
+  llvm::SmallVector<StringAttr> permutedDims;
+  for (auto dim : getRepOrder()) {
+    permutedDims.push_back(canonicalDims[dim]);
+    namedShape[canonicalDims[dim]] = shape[dim];
+  }
+  ll = ll.transposeOuts(permutedDims);
+  ll = ensureLayoutNotSmallerThan(ll, namedShape);
+  ll = ensureLayoutNotLargerThan(ll, namedShape);
+  ll = ll.transposeOuts(canonicalDims);
+  return ll;
+}
+
+SmallVector<unsigned>
+LinearEncodingAttr::getElemsPerThread(ArrayRef<int64_t> shape, Type) const {
+  // Broadcast to the shape and then count registers
+  auto ll = *toLinearLayout(shape);
+  return basesPerDim(ll, StringAttr::get(getContext(), "register"));
+}
+
+// Start of Selection
+SmallVector<unsigned> LinearEncodingAttr::getContigPerThread() const {
```

Comment:
I think this is similar to `getNumConsecutiveInOut`. Maybe we can remove one of the two methods